Core/Src/main.c

```diff
@@ -1,181 +1,216 @@
+  * License. You may obtain a copy of the License at:
+  *                        opensource.org/licenses/BSD-3-Clause
+  *
+  ******************************************************************************
+  */
+/* USER CODE END Header */
+/* Includes ------------------------------------------------------------------*/
+#include "main.h"
+#include "spi.h"
+#include "tim.h"
+#include "usart.h"
+#include "gpio.h"
+
+/* Private includes ----------------------------------------------------------*/
+/* USER CODE BEGIN Includes */
+
+/* USER CODE END Includes */
+
+/* Private typedef -----------------------------------------------------------*/
+/* USER CODE BEGIN PTD */
+
+/* USER CODE END PTD */
+
+/* Private define ------------------------------------------------------------*/
+/* USER CODE BEGIN PD */
+/* USER CODE END PD */
+
+/* Private macro -------------------------------------------------------------*/
+/* USER CODE BEGIN PM */
+
+/* USER CODE END PM */
+
+/* Private variables ---------------------------------------------------------*/
+
+/* USER CODE BEGIN PV */
+
+/* USER CODE END PV */
+
+/* Private function prototypes -----------------------------------------------*/
+void SystemClock_Config(void);
+/* USER CODE BEGIN PFP */
+
+/* USER CODE END PFP */
+
+/* Private user code ---------------------------------------------------------*/
+/* USER CODE BEGIN 0 */
+
+/* USER CODE END 0 */
+
+/**
+  * @brief  The application entry point.
+  * @retval int
+  */
+int main(void)
+{
+  /* USER CODE BEGIN 1 */
+
+  uint8_t tx_data[3] = {0x1, 0x80, 0x0}, rx_data[3];
+  uint16_t foo;
+
+  /* USER CODE END 1 */
+
+  /* MCU Configuration--------------------------------------------------------*/
+
+  /* Reset of all peripherals, Initializes the Flash interface and the Systick. */
+  HAL_Init();
+
+  /* USER CODE BEGIN Init */
+
+  /* USER CODE END Init */
+
+  /* Configure the system clock */
+  SystemClock_Config();
+
+  /* USER CODE BEGIN SysInit */
+
+  /* USER CODE END SysInit */
+
+  /* Initialize all configured peripherals */
+  MX_GPIO_Init();
+  MX_USART2_UART_Init();
+  MX_SPI1_Init();
+  MX_TIM1_Init();
+  /* USER CODE BEGIN 2 */
+
+  // send high to Chip Select ping
+  HAL_GPIO_WritePine(GPIOB, GPIO_PIN_8, GPIO_PIN_SET);
+
+  // start PWM signal generation
+  HAL_TIM_PWM_Start(&htim1, TIM_CHANNEL_1);
+
+  /* USER CODE END 2 */
+
+  /* Infinite loop */
+  /* USER CODE BEGIN WHILE */
+  while (1)
+  {
+
+	// send low to Chip Select pin to start reading
+	HAL_GPIO_WritePine(GPIOB, GPIO_PIN_8, GPIO_PIN_RESET);
+
+	// receive data from ADC
+	 hal_status = HAL_SPI_TransmiteReceive(&hspi1, tx_data, rx_data, 3, 100);
+
+	// send high to Chip Select pin to stop reading
+	HAL_GPIO_WritePine(GPIOB, GPIO_PIN_8, GPIO_PIN_SET);
+
+	 if (hal_status == HAL_OK){
+
+		// convert ADC value into number of counts for timer counter
+		foo = (((uint16_t) (rx_data[1] & 3)<<8) | rx_data[2]) * 3.125 + 3200;
+

```

Comment:
So this is the importance of not having magic numbers such as 3.125 and 3200... i have no idea what they mean so I am not sure what is going on here in terms of the calculation. I kind of see what your doing, but did you convert the adc value (0-1023) into a percentage? Or was this taken into account by multiplying by 3.125?